### PR TITLE
feat: add type_hint to use_state for typed state deserialization (#2198)

### LINKED ISSUE #2198: [FEATURE] allow users to specify type hints for component states
In #2015 we introduced component states, but since there's nowhere to specify a type hint, we're using the generic deserializer when loading value from the database.

We want to extend the API to allow an explicit type hint passed into `use_state()` call, so the value can be deserialized into a specific type (e.g. a specific dataclass).

---
❤️ Contributors, please refer to 📙[Contributing Guide](https://cocoindex.io/docs/about/contributing).
Unless the PR can be sent immediately (e.g. just a few lines of code), we recommend you to leave a comment on the issue like **`I'm working on it`**  or **`Can I work on this issue?`** to avoid duplicating work. Our [Discord server](https://discord.com/invite/zpA9S2DR7s) is always open and friendly.

### PR DESCRIPTION
**Pull request by lichuang**
Allow users to pass a `type_hint` to `coco.use_state()` so stored values are deserialized into a concrete type (dataclass, NamedTuple, msgspec.Struct, pydantic BaseModel) instead of the generic Any returned by the default decoder. This brings type-safe access to persistent component state, with IDE auto-complete and static type checking on `.value`.

Changes:
- `use_state()` gains an optional keyword-only `type_hint` parameter
- `StateHandle` holds a per-instance deserializer; `.value` delegates to it instead of the module-level `_DESERIALIZE_ANY` fallback
- New overload covers `use_state(key, initial_value, *, type_hint)` call pattern
- Deserialization errors include the state key via `source_label` for debugging
- 7 new tests: dataclass / NamedTuple / msgspec.Struct / pydantic round-trips, no-initial edge case, overload resolution, and type-mismatch error path

Closes #2198.

### CODE REVIEW

**Comment by badmonster0** [comment]
@georgeh0 can help take a look!

File: python/cocoindex/_internal/api.py [hunk 1]
```diff
@@ -736,7 +781,14 @@ def use_state(key: StableKey, initial_value: Any = None) -> StateHandle[Any]:
         # Rust client errors surface as ValueError; normalize to RuntimeError so
         # all use_state usage errors have a consistent type for callers.
         raise RuntimeError(str(e)) from None
-    return StateHandle(key, stored, ctx._core_processor_ctx)
+    if type_hint is not None:
+        deserializer = make_deserialize_fn(
+            type_hint,
+            source_label=f"use_state key {key!r}",
+        )
```
Comment (georgeh0) [inline python/cocoindex/_internal/api.py]: We may consider using this one (after renaming), which is reused rather than constructed each time on the fly: https://github.com/cocoindex-io/cocoindex/blob/4667e562979f6e8d0114c91b69bfbd5bf055f78c/python/cocoindex/_internal/serde.py#L461-L463

File: python/cocoindex/_internal/api.py [hunk 2]
```diff
@@ -676,8 +679,27 @@ def value(self, new_value: _StateT) -> None:
...
+@overload
 def use_state(key: StableKey, initial_value: _StateT) -> StateHandle[_StateT]: ...
-def use_state(key: StableKey, initial_value: Any = None) -> StateHandle[Any]:
+@overload
+def use_state(
+    key: StableKey,
+    *,
+    type_hint: type[_StateT],
+    initial_value: _StateT | None = None,
+) -> StateHandle[_StateT]: ...
```
Comment (georgeh0) [inline python/cocoindex/_internal/api.py:688]: minor: I think without `type_hint` we may just return `StateHandle[Any]` (there's no guarantee for the returned type). Maybe 2 overload variants are sufficient?

```python
@overload
def use_state(key: StableKey, initial_value: Any = None) -> StateHandle[Any]: ...
@overload
def use_state(
    key: StableKey,
    initial_value: _StateT | None = None,
    *,
    type_hint: type[_StateT],
) -> StateHandle[_StateT]:
  ...
```

**Review by georgeh0 — APPROVED** [review]
Thanks!
